#**MACHINE LEARNING PROJECT ON RED WINE DATASET**

# Classification

**Data preprocessing**

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [ ]:
dataset = pd.read_csv('/content/RED WINE.csv')

In [ ]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


In [ ]:
dataset['quality'].unique()

array([5, 6, 7, 4, 8, 3])

**Separate features and target**  


In [ ]:
X = dataset.iloc[: , 1:-1].values
y = dataset.iloc[:, -1].values

**Splitting data into train and test sets**

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X ,y, test_size=0.20)

**Standardization**

In [ ]:
from sklearn.preprocessing import StandardScaler

In [ ]:
sc = StandardScaler()

In [ ]:
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

**Train on different algorithms**

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC

In [ ]:
l_cla = LogisticRegression()
k_cla = KNeighborsClassifier()
d_cla = DecisionTreeClassifier()
r_cla = RandomForestClassifier()
s_cla = SVC(kernel='linear')
ks_cla = SVC(kernel= 'rbf')

In [ ]:
l_cla.fit(X_train, y_train)
k_cla.fit(X_train, y_train)
d_cla.fit(X_train, y_train)
r_cla.fit(X_train, y_train)
s_cla.fit(X_train, y_train)
ks_cla.fit(X_train, y_train)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [ ]:
l_pred = l_cla.predict(X_test)
k_pred = k_cla.predict(X_test)
d_pred = d_cla.predict(X_test)
r_pred = r_cla.predict(X_test)
s_pred = s_cla.predict(X_test)
ks_pred = ks_cla.predict(X_test)

In [ ]:
from sklearn.metrics import confusion_matrix

In [ ]:
l_c = confusion_matrix(y_test, l_pred)
k_c = confusion_matrix(y_test, k_pred)
d_c = confusion_matrix(y_test, d_pred)
r_c = confusion_matrix(y_test, r_pred)
s_c = confusion_matrix(y_test, s_pred)
ks_c = confusion_matrix(y_test, ks_pred)

In [ ]:
l_c

array([[ 0,  0,  3,  0,  0,  0],
       [ 0,  0, 11,  1,  1,  0],
       [ 0,  0, 97, 36,  3,  0],
       [ 0,  0, 44, 70, 15,  2],
       [ 0,  0,  4, 21, 10,  0],
       [ 0,  0,  0,  0,  2,  0]])

In [ ]:
k_c

array([[ 0,  1,  2,  0,  0,  0],
       [ 0,  0,  7,  6,  0,  0],
       [ 0,  1, 99, 33,  3,  0],
       [ 0,  0, 39, 82, 10,  0],
       [ 0,  0,  8, 17, 10,  0],
       [ 0,  0,  0,  0,  2,  0]])

In [ ]:
d_c

array([[ 1,  0,  1,  1,  0,  0],
       [ 1,  2,  6,  4,  0,  0],
       [ 0,  9, 97, 25,  4,  1],
       [ 0,  6, 37, 72, 11,  5],
       [ 0,  0,  3, 17, 15,  0],
       [ 0,  0,  0,  0,  2,  0]])

In [ ]:
r_c

array([[  0,   0,   3,   0,   0,   0],
       [  0,   0,   9,   4,   0,   0],
       [  0,   1, 106,  24,   5,   0],
       [  0,   1,  33,  85,  12,   0],
       [  0,   0,   2,  17,  16,   0],
       [  0,   0,   0,   0,   2,   0]])

In [ ]:
s_c

array([[  0,   0,   3,   0,   0,   0],
       [  0,   0,  11,   2,   0,   0],
       [  0,   0, 101,  34,   1,   0],
       [  0,   0,  48,  82,   1,   0],
       [  0,   0,   4,  31,   0,   0],
       [  0,   0,   0,   1,   1,   0]])

In [ ]:
ks_c

array([[  0,   0,   2,   1,   0,   0],
       [  0,   0,  10,   3,   0,   0],
       [  0,   0, 105,  30,   1,   0],
       [  0,   0,  43,  81,   7,   0],
       [  0,   0,   1,  24,  10,   0],
       [  0,   0,   0,   0,   2,   0]])

In [ ]:
from sklearn.metrics import accuracy_score

In [ ]:
l_a = accuracy_score(y_test, l_pred)
k_a = accuracy_score(y_test, k_pred)
d_a = accuracy_score(y_test, d_pred)
r_a = accuracy_score(y_test, r_pred)
s_a = accuracy_score(y_test, s_pred)
ks_a = accuracy_score(y_test, ks_pred)

**Result**

In [ ]:
print('Logistic Regression: ' + str(l_a) + '\nKNN: ' + str(k_a) + '\nDecision Tree: ' + str(d_a) + '\nRandom Forest: ' + str(r_a) + '\nLinear SVC: ' + str(s_a) + '\nKernel SVC: ' + str(l_a))

Logistic Regression: 0.553125
KNN: 0.596875
Decision Tree: 0.584375
Random Forest: 0.646875
Linear SVC: 0.571875
Kernel SVC: 0.553125


#Regression

**Data preprocessing**

In [ ]:
import pandas as pd
import pandas as pd
import matplotlib.pyplot as plt

In [ ]:
data=pd.read_csv("/content/RED WINE.csv")

In [ ]:
data

fixed acidity  volatile acidity  citric acid  ...  sulphates  alcohol  quality
0               7.4             0.700         0.00  ...       0.56      9.4        5
1               7.8             0.880         0.00  ...       0.68      9.8        5
2               7.8             0.760         0.04  ...       0.65      9.8        5
3              11.2             0.280         0.56  ...       0.58      9.8        6
4               7.4             0.700         0.00  ...       0.56      9.4        5
...             ...               ...          ...  ...        ...      ...      ...
1594            6.2             0.600         0.08  ...       0.58     10.5        5
1595            5.9             0.550         0.10  ...       0.76     11.2        6
1596            6.3             0.510         0.13  ...       0.75     11.0        6
1597            5.9             0.645         0.12  ...       0.71     10.2        5
1598            6.0             0.310         0.47  ...       0.66     11.0        6

[1599 rows x 12 columns]

In [ ]:
data. describe()

fixed acidity  volatile acidity  ...      alcohol      quality
count    1599.000000       1599.000000  ...  1599.000000  1599.000000
mean        8.319637          0.527821  ...    10.422983     5.636023
std         1.741096          0.179060  ...     1.065668     0.807569
min         4.600000          0.120000  ...     8.400000     3.000000
25%         7.100000          0.390000  ...     9.500000     5.000000
50%         7.900000          0.520000  ...    10.200000     6.000000
75%         9.200000          0.640000  ...    11.100000     6.000000
max        15.900000          1.580000  ...    14.900000     8.000000

[8 rows x 12 columns]

In [ ]:
data.head()

fixed acidity  volatile acidity  citric acid  ...  sulphates  alcohol  quality
0            7.4              0.70         0.00  ...       0.56      9.4        5
1            7.8              0.88         0.00  ...       0.68      9.8        5
2            7.8              0.76         0.04  ...       0.65      9.8        5
3           11.2              0.28         0.56  ...       0.58      9.8        6
4            7.4              0.70         0.00  ...       0.56      9.4        5

[5 rows x 12 columns]

In [ ]:
data. info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


**Separate features and target**

In [ ]:
X = data.iloc[: , :-1].values
Y = data.iloc[: , -1].values

In [ ]:
X

array([[ 7.4  ,  0.7  ,  0.   , ...,  3.51 ,  0.56 ,  9.4  ],
       [ 7.8  ,  0.88 ,  0.   , ...,  3.2  ,  0.68 ,  9.8  ],
       [ 7.8  ,  0.76 ,  0.04 , ...,  3.26 ,  0.65 ,  9.8  ],
       ...,
       [ 6.3  ,  0.51 ,  0.13 , ...,  3.42 ,  0.75 , 11.   ],
       [ 5.9  ,  0.645,  0.12 , ...,  3.57 ,  0.71 , 10.2  ],
       [ 6.   ,  0.31 ,  0.47 , ...,  3.39 ,  0.66 , 11.   ]])

In [ ]:
Y

array([5, 5, 5, ..., 6, 5, 6])

**Splitting Data into Train and Test sets**

In [ ]:
 from sklearn.model_selection import train_test_split

In [ ]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2)

In [ ]:
X_train

array([[7.60e+00, 6.30e-01, 3.00e-02, ..., 3.44e+00, 6.40e-01, 1.09e+01],
       [1.15e+01, 1.80e-01, 5.10e-01, ..., 3.28e+00, 9.70e-01, 1.01e+01],
       [8.00e+00, 3.10e-01, 4.50e-01, ..., 3.15e+00, 8.10e-01, 1.25e+01],
       ...,
       [1.19e+01, 3.70e-01, 6.90e-01, ..., 3.00e+00, 6.50e-01, 1.28e+01],
       [7.10e+00, 7.50e-01, 1.00e-02, ..., 3.39e+00, 4.00e-01, 1.28e+01],
       [8.50e+00, 6.60e-01, 2.00e-01, ..., 3.13e+00, 4.80e-01, 9.20e+00]])

**Train on different algorithms**

In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor

In [ ]:
m_reg = LinearRegression()
p_reg = LinearRegression()
d_reg = DecisionTreeRegressor()
r_reg = RandomForestRegressor(n_estimators=500)

In [ ]:
X_poly = PolynomialFeatures(degree =2)
X_poly = X_poly.fit_transform(X_train)

In [ ]:
m_reg.fit(X_train, Y_train)
p_reg.fit(X_poly,Y_train)
d_reg.fit(X_train, Y_train)
r_reg.fit(X_train, Y_train)

RandomForestRegressor(bootstrap=True, ccp_alpha=0.0, criterion='mse',
                      max_depth=None, max_features='auto', max_leaf_nodes=None,
                      max_samples=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      n_estimators=500, n_jobs=None, oob_score=False,
                      random_state=None, verbose=0, warm_start=False)

In [ ]:
temp = PolynomialFeatures(degree =2)
temp = temp.fit_transform(X_test)

In [ ]:
m_pred = m_reg.predict(X_test)
p_pred = p_reg.predict(temp)
d_pred = d_reg.predict(X_test)
r_pred = r_reg.predict(X_test)

In [ ]:
from sklearn.metrics import r2_score

In [ ]:
m = r2_score(Y_test, m_pred)
p = r2_score(Y_test, p_pred)
d = r2_score(Y_test, d_pred)
r = r2_score(Y_test, r_pred)

**Result**

In [ ]:
print('Logistic Regression: ' + str(m) + '\npolyfeat: ' + str(p) + '\nDecision Tree: ' + str(d) + '\nRandom Forest: ' + str(r))

Logistic Regression: 0.39192582636682416
polyfeat: 0.3444783096804007
Decision Tree: 0.17002012072434614
Random Forest: 0.5189738430583501


#**REPORT OF RED WINE DATASET**

##**1.Contents of data file**

Data file contains datasets of red_wine_quality. what are the elements needed to prepare the wine,relation between them.



##**2.Libraries imported**

i.Pandas  
ii.numpy  
iii.matplotlib

##**3.Model used**

i.Classification   
ii.Regression 

##**4.Algorithms**

1.Logistic regression  
2.KNN  
3.Decision tree  
4.Random forest  
5.Linear SVC  
6.Kernel SVC  
7.Polynomial features  


##**5.Data preprocessing**

1.describing data based on mean,median,std_dev,percentile   
2.finding null values if present  
3.checking datatypes  
4.scaling data  
5.encoding mechanisms  
6.checking quality of wine

##**6.Results**

two algorithms which have been used has given atmost equivalent results  
1.logistic regression 0.553125  
2.kernel SVC 0.553125 

Random forest gives best accuracy 0.646875

##**7.Conclusion**


according to given dataset,wine quality will be good if it cross the scale above 75%.

after testing on them the purity value obtained are less than given reference value.hence, the given the quality of wine is not good..

Since Random forest is giving more efficiency than other algorithm we can consider random forest as the best algorithm for this dataset